compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -788,23 +849,48 @@ mod llvm_enzyme {
         d_decl.inputs = d_inputs.into();
 
         if x.mode.is_fwd() {
+            let ty = match d_decl.output {
+                FnRetTy::Ty(ref ty) => ty.clone(),
+                FnRetTy::Default(span) => {
+                    // We want to return std::hint::black_box(()).
+                    let kind = TyKind::Tup(ThinVec::new());
+                    let ty = P(rustc_ast::Ty { kind, id: ast::DUMMY_NODE_ID, span, tokens: None });
+                    d_decl.output = FnRetTy::Ty(ty.clone());
+                    assert!(matches!(x.ret_activity, DiffActivity::None));
+                    // this won't be used below, so any type would be fine.
+                    ty
+                }
+            };
+
             if let DiffActivity::Dual = x.ret_activity {
-                let ty = match d_decl.output {
-                    FnRetTy::Ty(ref ty) => ty.clone(),
-                    FnRetTy::Default(span) => {
-                        panic!("Did not expect Default ret ty: {:?}", span);
-                    }
+                let kind = if x.width == 1 {
+                    // Dual can only be used for f32/f64 ret.
+                    // In that case we return now a tuple with two floats.
+                    TyKind::Tup(thin_vec![ty.clone(), ty.clone()])
```

Comment:
wouldn't it be easier to normalize everything to using arrays and not having to have all these `if width == 1` or related checks?

Reply:
Especially since Rust does not guarantee an order of tuple fields in memory

Reply:
The order sounds like a problem, but all the "wiring up" of args happens on llvm-ir, after enzyme generated the function, so I don't think rust has anything to say. After all we delete the whole dummy function which rust lowered to llvm-ir anyway. 

Normalizing it always to arrays would be nice for us, but also be annoying for users.
I am guessing that 3/4 will use scalar mode autodiff and 1/4 would use batch mode autodiff.
So in 3/4 of the cases we would then return `[f32;1]` and everyone would be writing 
`let x = df(some,args)[0];`

Reply:
Oh lmao. This user facing 🤦